manager/volume.go

```diff
@@ -350,26 +397,66 @@ func (m *VolumeManager) Detach(name string) (v *longhorn.Volume, err error) {
 	if err != nil {
 		return nil, err
 	}
-	if v.Status.State != types.VolumeStateAttached && v.Status.State != types.VolumeStateAttaching {
-		return nil, fmt.Errorf("invalid state to detach %v: %v", v.Name, v.Status.State)
+
+	if v.Status.IsStandby {
+		return nil, fmt.Errorf("cannot detach standby volume %v", v.Name)
 	}
 
-	oldNodeID := v.Spec.NodeID
-	if oldNodeID == "" {
+	if v.Spec.NodeID == "" && v.Spec.MigrationNodeID == "" {
+		logrus.Debugf("No need to detach volume %v is already detached from all nodes", v.Name)
 		return v, nil
 	}
 
-	v.Spec.NodeID = ""
-	v.Spec.DisableFrontend = false
+	if nodeID != "" && nodeID != v.Spec.NodeID && nodeID != v.Spec.MigrationNodeID {
+		logrus.Debugf("No need to detach volume %v since it's not attached to node %v", v.Name, nodeID)
+		return v, nil
+	}
+
+	// if the detach request doesn't specify a node to detach from
+	// we detach from all nodes, which is the same behavior as we previously had
+	if nodeID == "" {
+		v.Spec.NodeID = ""
+		v.Spec.MigrationNodeID = ""
+		logrus.Infof("Volume %v detachment from all nodes requested", v.Name)
+	} else if nodeID == v.Spec.NodeID {
+		if v.Spec.MigrationNodeID != "" && v.Spec.Migratable {
+			// migration confirmation, since detach is requested from the old node
+			// the engine must be running on the new node in order to be confirmed
+			if !m.isVolumeAvailableOnNode(name, v.Spec.MigrationNodeID) {
+				return nil, fmt.Errorf("migration is not ready yet")
+			}
+			v.Spec.NodeID = v.Spec.MigrationNodeID
+			v.Spec.MigrationNodeID = ""
+			logrus.Infof("Volume %v migration from %v to %v confirmed", v.Name, nodeID, v.Spec.NodeID)
+		} else {
+			v.Spec.NodeID = ""
+			v.Spec.MigrationNodeID = ""
+			logrus.Infof("Volume %v detachment from node %v requested", v.Name, nodeID)
+		}
+	} else if nodeID == v.Spec.MigrationNodeID {
+		v.Spec.MigrationNodeID = ""
+		logrus.Infof("Volume %v migration from %v to %v rollback", v.Name, nodeID, v.Spec.NodeID)
+	}
```

Comment:
```
isMigratingVolume := v.Spec.Migratable && v.Spec.MigrationNodeID != "" && v.Spec.NodeID != ""
isMigrationConfirmation := isMigratingVolume && nodeID == v.Spec.NodeID
isMigrationRollback := isMigratingVolume && nodeID == v.Spec.MigrationNodeID

// Since all invalid/unexcepted cases have been handled above, we only need to take care of regular detach or migration here.
if isMigrationConfirmation {
} else if isMigrationRollback {
} else {  // Regular detach
}
```

Reply:
Thanks makes it simpler.